httpcore5/src/main/java/org/apache/hc/core5/http/ssl/TLS.java

```diff
@@ -0,0 +1,97 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * ====================================================================
+ *
+ * This software consists of voluntary contributions made by many
+ * individuals on behalf of the Apache Software Foundation.  For more
+ * information on the Apache Software Foundation, please see
+ * <http://www.apache.org/>.
+ *
+ */
+
+package org.apache.hc.core5.http.ssl;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.apache.hc.core5.http.ParseException;
+import org.apache.hc.core5.http.ProtocolVersion;
+import org.apache.hc.core5.http.message.ParserCursor;
+
+/**
+ * Supported {@code TLS} protocol versions.
+ *
+ * @since 5.0
+ */
+public enum TLS {
+
+    V_1_0("TLSv1",   new ProtocolVersion("TLS", 1, 0)),
+    V_1_1("TLSv1.1", new ProtocolVersion("TLS", 1, 1)),
+    V_1_2("TLSv1.2", new ProtocolVersion("TLS", 1, 2)),
+    V_1_3("TLSv1.3", new ProtocolVersion("TLS", 1, 3));
+
+    public final String ident;
+    public final ProtocolVersion version;
+
+    TLS(final String ident, final ProtocolVersion version) {
+        this.ident = ident;
+        this.version = version;
+    }
+
+    public boolean isSame(final ProtocolVersion protocolVersion) {
+        return version.equals(protocolVersion);
+    }
+
+    public boolean isComparable(final ProtocolVersion protocolVersion) {
+        return version.isComparable(protocolVersion);
+    }
+
+    public boolean greaterEquals(final ProtocolVersion protocolVersion) {
+        return version.greaterEquals(protocolVersion);
+    }
+
+    public boolean lessEquals(final ProtocolVersion protocolVersion) {
+        return version.lessEquals(protocolVersion);
+    }
+
+    public static ProtocolVersion parse(final String s) throws ParseException {
+        if (s == null) {
+            return null;
+        }
+        final ParserCursor cursor = new ParserCursor(0, s.length());
+        return TlsVersionParser.INSTANCE.parse(s, cursor, null);
+    }
+
+    public static String[] excludeWeak(final String... protocols) {
+        if (protocols == null) {
+            return null;
+        }
+        final List<String> enabledProtocols = new ArrayList<>();
+        for (final String protocol: protocols) {
+            if (!protocol.startsWith("SSL") && !protocol.equals(V_1_0.ident) && !protocol.equals(V_1_1.ident)) {
+                enabledProtocols.add(protocol);
+            }
+        }
+        if (enabledProtocols.isEmpty()) {
+            enabledProtocols.add(V_1_2.ident);
```

Comment:
Technically, there may be newer protocol versions available such as TLS 1.3 which may be considered as a better option. I am not sure if the method should figure out what is the latest protocol version available but maybe it should just return an empty array here. What do you think?

Reply:
@artem-smotrakov The trouble is that if  `enabledProtocols` is set to `null` that will prompt some JSSE provides to fall back onto the default set of protocols which may contain the same weak protocols we have just filtered out. We need to provide at least one value and presently `TLSv12` looks to be the most widely used strong protocol.

Reply:
I see. TLSv12 seems to be a good choice. Higher protocol versions may be better but the method should be smarter to pick up a better option.